sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -449,37 +449,109 @@ def test_spline_transformer_n_features_out(n_knots, include_bias, degree):
-                    x1 ** 0 * x2 ** 2])
-    deg2 = 2
+    with pytest.raises(ValueError, match=err_msg):
+        PolynomialFeatures(**params).fit(X)
+
+
+@pytest.fixture()
+def single_feature_degree2():
+    X = np.arange(6)[:, np.newaxis]
+    P = np.hstack([np.ones_like(X), X, X ** 2, X ** 3])
+    return X, P
 
-    for (deg, X, P) in [(deg1, X1, P1), (deg2, X2, P2)]:
-        P_test = PolynomialFeatures(deg, include_bias=True).fit_transform(X)
-        assert_array_almost_equal(P_test, P)
 
-        P_test = PolynomialFeatures(deg, include_bias=False).fit_transform(X)
-        assert_array_almost_equal(P_test, P[:, 1:])
+@pytest.mark.parametrize(
+    "degree, include_bias, interaction_only, indices",
+    [
+        (3, True, False, slice(None, None)),
+        (3, False, False, slice(1, None)),
+        (3, True, True, [0, 1]),
+        (3, False, True, [1]),
+        ((2, 3), True, False, [0, 2, 3]),
+        ((2, 3), False, False, [2, 3]),
+        ((2, 3), True, True, [0]),
+        ((2, 3), False, True, []),
+    ]
+)
+def test_polynomial_features_degree_3(
+    single_feature_degree2,
+    degree,
+    include_bias,
+    interaction_only,
+    indices
+):
+    """Test Polynomial Features of degree 3."""
+    X, P = single_feature_degree2
+    tf = PolynomialFeatures(
+        degree=degree,
+        include_bias=include_bias,
+        interaction_only=interaction_only
+    ).fit(X)
+    assert_allclose(tf.transform(X), P[:, indices])
+    if tf.n_output_features_ > 0:
+        assert tf.powers_.shape == (tf.n_output_features_, tf.n_features_in_)
+
+
+@pytest.fixture()
+def two_feature_degree3():
+    X = np.arange(6).reshape((3, 2))
+    x1 = X[:, :1]
+    x2 = X[:, 1:]
+    P = np.hstack([x1 ** 0 * x2 ** 0,   # 0
+                   x1 ** 1 * x2 ** 0,   # 1
+                   x1 ** 0 * x2 ** 1,   # 2
+                   x1 ** 2 * x2 ** 0,   # 3
+                   x1 ** 1 * x2 ** 1,   # 4
+                   x1 ** 0 * x2 ** 2])  # 5
+    return X, P
 
-    interact = PolynomialFeatures(2, interaction_only=True, include_bias=True)
-    X_poly = interact.fit_transform(X)
-    assert_array_almost_equal(X_poly, P2[:, [0, 1, 2, 4]])
 
-    assert interact.powers_.shape == (interact.n_output_features_,
-                                      interact.n_input_features_)
+@pytest.mark.parametrize(
+    "degree, include_bias, interaction_only, indices",
+    [
+        (2, True, False, slice(None, None)),
+        (2, False, False, slice(1, None)),
+        (2, True, True, [0, 1, 2, 4]),
+        (2, False, True, [1, 2, 4]),
+        ((2, 2), True, False, [0, 3, 4, 5]),
+        ((2, 2), False, False, [3, 4, 5]),
+        ((2, 2), True, True, [0, 4]),
+        ((2, 2), False, True, [4]),
+    ]
+)
+def test_polynomial_features_degree_2(
+    two_feature_degree3,
+    degree,
+    include_bias,
+    interaction_only,
+    indices
+):
+    """Test Polynomial Features of degree 2."""
+    X, P = two_feature_degree3
```

Comment:
I think this should be parameterized to also test the case where X is a sparse matrix.

Reply:
I think I discovered an untested edge case where the transform(X) is an empty sparse csr matrix.